DM requests for invoices

benthecarman opened the pull request:
Think this is all the functionality needed for @futurepaul to start

TonyGiorgio commented on mutiny-core/src/lib.rs:
```diff
@@ -629,7 +629,7 @@ impl<S: MutinyStorage> MutinyWalletBuilder<S> {
     xprivkey: ExtendedPrivKey,
     config: MutinyWalletConfig,
-    pub(crate) storage: S,
+    pub storage: S,
```
Pass storage to things that need storage.

TonyGiorgio commented on mutiny-core/src/storage.rs:
```diff
@@ -268,6 +273,28 @@ pub trait MutinyStorage: Clone + Sized + Send + Sync + 'static {
     }
 
+    /// Insert a nsec into the storage
+    async fn insert_nsec(&self, nsec: SecretKey) -> Result<(), MutinyError> {
```
Hard NACK. Nsec should not be pasted into a browser site.

benthecarman replied:
Should discuss IRL but paul and I's thinking was we already doing worse with a bitcoin seed

TonyGiorgio replied:
Huge difference in a mutiny seed being used for mutiny things and importing another seed/private key from somewhere else for other purposes. 

Find another way, there's others that exist. Maybe ask Pablo...

TonyGiorgio replied:
There's a few things I think we can do here but I would absolutely not allow nsec pasting here. 

1. Nostr plugins for web browsers that already have them
2. Create an nsec from the derivation path and use that for users that don't already have nostr
  2a. Some delegated nostr stuff so you can use the mutiny derivation seed as if it were the user's main key (don't think this really exists yet)
3. nsec bunker for everyone else

benthecarman commented:
Changed to not import but be instead be part of the building phase. We can then instead store it somewhere more secure

TonyGiorgio commented:
Needs clippy fixes

TonyGiorgio commented on mutiny-core/src/nostr/mod.rs:
```diff
@@ -91,14 +106,17 @@ impl<S: MutinyStorage> NostrManager<S> {
             .collect();
 
+        // add relay to pull DMs from
+        relays.push("wss://relay.primal.net".to_string());
```
add damus

TonyGiorgio commented on mutiny-core/src/storage.rs:
```diff
@@ -27,6 +27,7 @@ use uuid::Uuid;
 
 pub const KEYCHAIN_STORE_KEY: &str = "bdk_keychain";
 pub const MNEMONIC_KEY: &str = "mnemonic";
+pub const NSEC_KEY: &str = "nsec";
```
delete this

benthecarman commented on mutiny-core/src/lib.rs:
```diff
@@ -1420,6 +1462,64 @@ impl<S: MutinyStorage> MutinyWallet<S> {
+        limit: u64,
+        until: Option<u64>,
+    ) -> Result<Vec<DirectMessage>, MutinyError> {
+        let url = primal_url.unwrap_or("https://primal-cache.mutinywallet.com/api");
```
Currently we have it an an ENV in the front end and pass it for the individual requests, probably could refactor to have this as part of init, maybe good for a follow up so we can do it with the sync nostr contacts stuff too

TonyGiorgio replied:
We're already changing the mutiny wallet init parameters, might as well add it in this PR

benthecarman commented on mutiny-core/src/nostr/mod.rs:
```diff
@@ -1011,14 +1183,33 @@ impl<S: MutinyStorage> NostrManager<S> {
+        // use provided nsec, otherwise generate it from seed
+        let (primary_key, public_key) = match key_source {
+            NostrKeySource::Derived => {
+                let keys = Self::derive_nostr_key(&context, xprivkey, 0, None, None)?;
```
Yeah we're using `0` to comply with nip6, will make it an const

https://github.com/nostr-protocol/nips/blob/master/06.md

benthecarman commented:
Was missing send functionality and an efficient way to get future messages. Added that. Cleaned up some of our nostr client handling as well so we don't need to make so many new connections

TonyGiorgio commented:
Tests are panicing
